### PR TITLE
fix(layer): mirror anchor-positioned popovers in RTL

### LINKED ISSUE #3389: [Bug] Dropdown broken if RTL
### Description

in the RTL environment the Dropdown component is broken

Image

### Reproduction

1. Render `<DropdownMenu />` or any `<MoreMenu /`
2. set parent element css `direction: rtl`
3. click on dropdown - the menu appears far from control

### XDS Version

v1.2.0

### Environment

_No response_

### PR DESCRIPTION
**Pull request by AKnassa**
## What this does
Makes dropdown menus and every other popover open on the correct side for right-to-left languages. Before, a menu in an RTL page opened as if the page were left-to-right — and on some browser versions it landed far away from its button entirely.

## Why
Anyone using the design system in an RTL context got menus growing the wrong way or detached from their trigger. Since all popovers share one positioning engine, this was systemic — not just DropdownMenu.

Fixes #3389

## What changed

1-before-rtl-wide-menu
2-after-rtl-wide-menu
3-ltr-unchanged
4-after-rtl-story

- The shared positioning engine (`useLayer`) now reads the trigger's text direction when a popover opens and mirrors its position accordingly. Works for both ways of setting direction (the `dir` attribute and the CSS `direction` property — the latter is what this issue's reproduction uses).
- The direction-resolved styles are committed synchronously before the popover is promoted to the top layer, so even the very first open paints on the correct side (review feedback — previously the first open could use the prior render's mapping for one frame).
- An explicit alignment is also set in RTL as a defense for browser versions that mis-handle the implied alignment — that's the "menu appears far from control" case in the report.
- New `positioning: 'custom'` opt-out on `useLayer`'s context render props for consumers that author their own position styles. Carousel (centered `anchor-size()` cover) and Tokenizer (explicit `anchor()` insets) use it instead of nulling out the hook's emitted styles from the outside (review feedback). `renderTooltip`/`renderHoverCard` exclude the option since those wrappers always derive their position.
- One fix covers all ~20 popover consumers: DropdownMenu, MoreMenu, Popover, Selector, MultiSelector, Typeahead, DateInput/DateTimeInput/DateRangeInput, SideNav flyouts, TopNav menus, Tooltip/HoverCard (when given side placements), and more.
- Left-to-right behavior is completely unchanged (existing placement tests pass unmodified; live measurements are byte-identical).
- 19 new tests: the RTL matrix, edge cases (direction flips between opens, fixed-mode immunity, indeterminate direction fallback, no-trigger safety, controlled opens), first-paint correctness asserted inside a mocked `showPopover()`, and the custom-positioning opt-out. Plus a new RTL Storybook story, docs updated in English/Chinese/dense variants, and changesets.

## Root cause (the deeper detail)
CSS `position-area` only supports **physical** keywords reliably today — the logical/self keyword families aren't implemented in current Chrome, and an invalid `position-area` pins the popover to the viewport corner. So `useLayer` maps its logical `placement`/`alignment` API to physical keywords itself, resolved against the trigger's computed direction at open time. That mapping, and the rationale for it, stays isolated inside Layer — consumers only ever see the logical API, and the physical fallback can be deleted in one place once logical keywords ship.

Separately, some engine versions resolve `position-area`'s **implied** alignment against the popover's own inherited direction instead of its containing block — that's how the reported menu landed at the viewport mirror of its trigger. Emitting an explicit physical `justify-self` (RTL only) removes that direction-sensitive resolution entirely; the `flip-inline` try tactic value-flips it, so fallback positions still hug the anchor.

## How to see it
Run Storybook and open **Core/DropdownMenu → RTL**: both menus (one per direction mechanism) open right-aligned to their trigger and grow left. Compare with any other DropdownMenu story for the unchanged LTR behavior. Screenshots attached below.

## Good to know (scope)
- The reported "far from control" displacement does not reproduce on current Chrome (it conforms to spec here); the explicit-alignment part of the fix defends against the engine class that shows it. The issue's Environment field is empty — reporter's browser/version would be useful to confirm.
- Known follow-ups deliberately not in this PR: Tokenizer authors its own position with a physical `left: anchor(start)` (via the new custom-positioning opt-out — still RTL-wrong, now explicitly the consumer's responsibility and marked as a follow-up in the source), and side-placement entry animations still slide from the physical side (cosmetic).

### CODE REVIEW

File: packages/core/src/Carousel/Carousel.tsx [hunk 1]
```diff
@@ -320,6 +320,9 @@ export function Carousel({
 
   const coverStyle: React.CSSProperties = {
     positionArea: 'center',
+    // Null the RTL justify-self useLayer may emit — the anchor-size cover
+    // must stay centered on the anchor with no self-alignment offset.
+    justifySelf: undefined,
```
Comment (cixzhang) [inline packages/core/src/Carousel/Carousel.tsx]: I guess this relies on knowing the internals for how useLayer applies RTL positioning under the hood. It may be easier to read with a dedicated option on the hook to disable rtl for these situations

**Review by cixzhang — CHANGES_REQUESTED** [review]
Thanks for digging into the RTL positioning issue — mirroring the shared Layer behavior seems like the right place to solve this. I’d like a few changes before merge:

- `useLayer.show()` reads direction with `setIsRtl(...)` and then immediately calls `showPopover()`, so the first open can still use the previous render’s `positionArea` / `justifySelf`. Could we make the RTL positioning apply synchronously before `showPopover()`, and add a test that asserts the style inside a mocked `showPopover` call?
- The `justifySelf: undefined` overrides in Carousel/Tokenizer make Layer internals leak into consumers. Could Layer suppress its auto RTL `justifySelf` when a consumer overrides `positionArea` / uses custom positioning? Carousel’s centered cover also looks like it shouldn’t need an override.
- This makes `start`/`end` logical in RTL, which is likely the right semantic, but it is consumer-visible. Please call that out clearly in the docs/changeset; if logical CSS anchor keywords aren’t usable yet, keeping the physical fallback isolated in Layer with that rationale would help.
- Please trim the changeset to a short release note. The root-cause detail is useful in the PR body, but the changelog entry should stay concise.

**Comment by AKnassa** [comment]
Thanks for the thorough review @cixzhang — all four points addressed:

**1. First-open race (`setIsRtl` → `showPopover()`):** Confirmed and fixed. `render()` now stashes the position props in a ref, and `show()` re-derives the direction and writes the direction-dependent styles (`position-area`, `justify-self`) synchronously onto the popover element before `showPopover()`, so the first paint — and any `@starting-style` entry animation — is direction-correct. I used direct pre-show application rather than `flushSync` because `show()` is called from passive effects (Carousel's mount effect, Selector's `isDefaultOpen`), where `flushSync` warns and silently degrades back to the async behavior. Two details the synchronous path is careful about: consumer-authored `style` keys are never touched — guarded by key presence, so the explicit-`undefined` nulling idiom keeps working (the documented merge order stands) — and the stash is promoted at commit time (layout effect), so `show()` only ever applies values from committed renders, never from a discarded concurrent pass. New tests assert the style inside a mocked `showPopover()`: first RTL open, reopen after an rtl→ltr flip, consumer-override safety, and the nulling idiom.

**2. `justifySelf` overrides:** Added `positioning: 'custom'` to the context render props — it keeps the popover behavior and `position-anchor` wiring but derives no position styles at all (`placement`/`alignment` ignored, RTL adjustments included). Carousel's centered cover and Tokenizer's inset-pinned layer now author their position entirely on their side; no more knowledge of hook internals, and you were right that Carousel needs no override at all once the derivation is suppressed at the source. `renderTooltip`/`renderHoverCard` exclude the option via `Omit` since those wrappers always derive their position (accepting it would be a silent no-op).

**3. Logical-semantics visibility:** The changeset now leads with the consumer-visible semantic ("`placement`/`alignment` `start`/`end` are logical — resolved against the trigger's computed direction at open time"), and the prop docs across Popover/Tooltip/HoverCard/useLayer (en/dense/zh) say the same. The physical keyword mapping stays isolated in `getPositionArea` inside Layer, with the rationale in-code (the logical `position-area` keyword families aren't implemented in Chrome yet, and an invalid value pins the popover to the viewport corner because the base styles zero the UA margins); the PR body now carries that rationale too, so the fallback can be deleted in one place once logical keywords ship.

**4. Changeset** trimmed to a short release note; the root-cause detail moved to the PR body.

File: packages/core/src/Carousel/Carousel.tsx (hunk 1 above)
Comment (AKnassa) [reply packages/core/src/Carousel/Carousel.tsx]: Done — added `positioning: 'custom'` to the context render props: it keeps the popover behavior and `position-anchor` wiring but derives no position styles (`placement`/`alignment` ignored, RTL adjustments included), so Carousel and Tokenizer now author their position entirely on their side instead of nulling the hook's emitted styles. Carousel's centered cover needed no override at all once the derivation is suppressed at the source.

File: packages/core/src/Layer/useLayer.tsx [hunk 2]
```diff
@@ -65,7 +69,32 @@ export type LayerAlignment = 'start' | 'center' | 'end';
...
+   * `'custom'`: the consumer authors its own position styles via `style`
+   * (e.g. explicit `anchor()` insets or an `anchor-size()` cover). The hook
+   * keeps the popover behavior and the `position-anchor` wiring but emits no
+   * placement-derived styles — including the RTL adjustments, so direction
+   * handling becomes the consumer's responsibility. `placement`/`alignment`
+   * are ignored.
+   *
+   * @default 'anchor'
+   */
+  positioning?: 'anchor' | 'custom';
```
Comment (cixzhang) [inline packages/core/src/Layer/useLayer.tsx:86]: Thanks for updating!

Thinking about this part a bit, it can be confusing to say that we have custom positioning apart from anchor positioning. Technically all positioning is anchored, but with custom we break free from the conventions from placement and alignement. I wonder if we need a new prop though? What if we allow `none` as an option for both `placement` and `alignment`? This would also avoid applying the styles for RTL in the specific cases we want to avoid it.

File: packages/core/src/Carousel/Carousel.tsx [hunk 3]
```diff
@@ -397,8 +399,7 @@ export function Carousel({
             </div>
           </>,
           {
-            placement: 'below',
-            alignment: 'center',
+            positioning: 'custom',
```
Comment (cixzhang) [inline packages/core/src/Carousel/Carousel.tsx:402]: Hmm using `below` and `center` placement shouldn't result in the extra RTL styling right? Ideally we should be able to keep the original placement and alignment here if it's unaffected.

**Review by cixzhang — COMMENTED** [review]
Thanks for working on this some more. I think I want to explore RTL on layers a bit since the proposal here is adds quite a lot of complexity and state management into the layer system.

**Review by cixzhang — CHANGES_REQUESTED** [review]
Thanks for the thorough writeup — the `positioning: 'custom'` opt-out and the consumer cleanups (Carousel/Tokenizer authoring their own styles instead of nulling from outside) are genuinely good and I'd keep those regardless.

My concern is with the core mechanism. The PR is built on this premise (from `getPositionArea`'s comment):

> position-area only supports physical keywords reliably (the logical/self keyword families are unsupported in current Chrome, and an invalid position-area pins the popover to the viewport corner)

I don't think that holds anymore, and if it doesn't, almost all of the new runtime complexity here is unnecessary. `position-area` **does** implement the logical `self-*` keyword family in current Chrome, and because we render the layer *inside the trigger's subtree*, it inherits `direction` — so `self-inline-start` / `self-inline-end` resolve against the popover's own direction and mirror in RTL **with zero JS**.

I verified the full 12-cell `placement` × `alignment` matrix in Chrome 150 (real `getBoundingClientRect`, both `dir` attribute and CSS `direction`):

- Physical mapping (`left`/`right`): identical in LTR and RTL — never mirrors (the bug).
- `self-*` logical mapping: **identical to physical in LTR, correctly mirrored in RTL**, every cell valid, no corner fallback.
- `flip-inline` fallbacks still work with `self-*` keywords (forced an overflow, it flipped correctly).

I suspect the "logical keywords don't work" conclusion came from trying the **plain** logical family (`inline-start`, no `self-`). Those resolve against the *containing block*, which for a top-layer popover is the page root — so they only flip when `dir` is on `<html>`, not on a subtree. That's exactly #3389's repro (`direction: rtl` on a parent), so plain logical looks broken. `self-*` is the fix: it resolves against the element's own inherited direction.

### Suggested change

Make `getPositionArea` emit `self-*` logical keywords and drop the direction handling entirely:

```ts
/**
 * Map logical placement/alignment to a CSS position-area value.
 *
 * Uses the self-* logical keyword family: the inline axis resolves against
 * the popover's own inherited direction (the layer renders inside the
 * trigger's subtree, so it inherits `direction` and mirrors in RTL with no
 * JS). The block axis is direction-neutral.
 */
function getPositionArea(
  placement: LayerPlacement = 'above',
  alignment: LayerAlignment = 'center',
): string {
  if (placement === 'above' || placement === 'below') {
    const block = placement === 'above' ? 'self-block-start' : 'self-block-end';
    if (alignment === 'start') return `${block} span-self-inline-end`;
    if (alignment === 'end') return `${block} span-self-inline-start`;
    return block; // center
  }
  const inline = placement === 'start' ? 'self-inline-start' : 'self-inline-end';
  if (alignment === 'start') return `${inline} span-self-block-end`;
  if (alignment === 'end') return `${inline} span-self-block-start`;
  return inline; // center
}
```

Note the two axes must come from the same keyword family — mixing physical `top` with `self-inline-*` produces an invalid `position-area` (computes to `none`). That's why the block axis moves to `self-block-*` too; in horizontal writing mode it's identical to `top`/`bottom`, and it's verified LTR-identical to today's output.

### Cleanup checklist (what this lets you delete)

Once the mapping is static, the entire show-time direction pipeline is dead code:

1. **`getPositionArea`** — remove the `isRtl` parameter and all `isRtl ? …` ternaries (replace body with the above).
2. **`getRtlJustifySelf`** — delete the whole function. With `self-*` keywords the inline alignment is spec-resolved against the popover's own direction, so the `justify-self` "defense" is no longer needed. (It was guarding against an engine class the PR itself notes it couldn't reproduce — if you still want a defense, it should come with an actual failing repro rather than carried as speculative complexity.)
3. **`const [isRtl, setIsRtl] = useState(false)`** — delete, plus every read/write.
4. **`show()`** — remove the entire `if (mode === 'context') { … }` block (the `getComputedStyle(triggerRef.current).direction` read, `setIsRtl`, and the synchronous `popover.style.positionArea` / `justifySelf` stamping). Revert `show()` to just the `showPopover` feature-check. `mode` can come out of its dep array again.
5. **The dual-ref stash** — delete `ContextPositionStash`, `pendingContextPositionRef`, `contextPositionRef`, the promotion line in `popoverRefCallback`, and the `useLayoutEffect(() => { contextPositionRef.current = pendingContextPositionRef.current; })`. All of it exists only to feed committed values into the show-time stamping; with no stamping it's unreachable. Drop the `useLayoutEffect` import if now unused.
6. **`renderContext`** — remove the `rtlJustifySelf` computation and the `...(rtlJustifySelf != null && {justifySelf})` spread, and the `pendingContextPositionRef.current = {…}` write. Remove `isRtl` from the `useCallback` dep array. Keep the `positioning === 'custom'` branch and `positionTryFallbacks` unchanged.

Keep: the `positioning: 'anchor' | 'custom'` render prop and its `custom` branch, and `positionTryFallbacks` (works with `self-*`).

### Tests & docs

- The current tests run in **jsdom**, which has no layout engine and doesn't inherit `direction`, so they only assert the emitted `position-area` *string* — they can't actually verify mirroring. The show-time/`getComputedStyle`, first-paint-inside-`showPopover`, and discarded-render/dual-ref tests all become moot. Replace with a simple table asserting `getPositionArea(placement, alignment)` returns the expected `self-*` string (LTR and RTL are now the same string by construction). Tests asserting literal `position-area: bottom` etc. will need updating to the new `self-block-*` strings.
- Please keep the **RTL Storybook story** — with jsdom unable to validate geometry, that (plus a browser check) is the real proof surface.
- Update `useLayer.doc.mjs` and the SYNC/JSDoc to describe the `self-*` logical mapping.

### Changeset

The RTL changeset says "resolved against the trigger's computed direction at open time," which won't be true anymore. Something like: `start`/`end` are logical (inline-start/inline-end) and popovers mirror automatically in RTL via CSS — keep it to the one-line user-facing effect.

### One caveat

My evidence is Chrome 150. The only legitimate reason to keep the physical mapping would be a **browser-support window** — if `self-*` keywords shipped meaningfully later than base anchor positioning (Chrome 125) and we must support that gap. I couldn't find a `browserslist` floor in the repo, and the code already assumes the Popover API + anchor positioning, and the PR's own `flip-block`/`flip-inline` fallbacks are the same vintage. If there *is* an explicit old-engine requirement I'm missing, that would change the calculus — worth confirming either way.

Requesting changes to swap to the logical mapping and strip the runtime machinery. Happy to pair on it if useful.

**Comment by cixzhang** [comment]
Addendum / correction to my review above — on reflection, **also remove `positioning: 'custom'`**. I'd flagged it as "keep regardless," but it's net-new public API for what is really two internal consumers, and the pre-existing nulling idiom on `main` already covers the same need without expanding the hook's surface. Given the goal here is to *shrink* Layer's complexity, this option cuts the other way.

### Remove the custom-positioning opt-out

7. **`ContextRenderProps`** — delete the `positioning?: 'anchor' | 'custom'` field and its JSDoc; drop the `positioning = 'anchor'` destructure and the `positioning === 'custom'` branch in `renderContext` (always emit the derived `anchorStyle`).
8. **`useTooltip` / `useHoverCard`** — revert the `Omit<ContextRenderProps, 'positioning'>` signatures back to `ContextRenderProps` (both the type and the callback params). Those `Omit`s only exist to hide the new option.
9. **Carousel** and **Tokenizer** — revert their `render(...)` calls and style objects to the `origin/main` form. On `main` both suppress the hook's derived styles from the outside via the style object (`coverStyle` overrides `positionArea: 'center'`; `popoverOverrideStyle` sets `positionArea: undefined` / `positionTryFallbacks: undefined` alongside its explicit `anchor()` insets). That idiom still works — the consumer `style` is spread last over `anchorStyle`, so an explicit `undefined` clears the derived value. Restore `placement`/`alignment` on those calls as they were.
10. **Changeset** — drop `.changeset/layer-custom-positioning.md` entirely (no public API is being added anymore).

Net effect: Layer goes back to a single positioning path (derive from `placement`/`alignment`), and the only real change in the PR becomes the `self-*` mapping in `getPositionArea`.

### Separate, unrelated: please don't delete `htmlName`

While tracing the Tokenizer changes I noticed this PR also **removes the `htmlName` prop and its hidden-input form participation** (the `htmlName?: string` prop, the destructure, and the `value.map(... <input type="hidden">)` block), plus its tests. That API is on `main` (added in #3558) and is unrelated to the RTL fix — this looks like a bad rebase/merge dropping it. Please restore it (rebase on `main` and bring only the RTL change) so the PR doesn't silently revert a shipped feature. This is the main reason I'm keeping the review at *request changes* independent of the mapping discussion.

**Comment by AKnassa** [comment]
Done in baaaf12 — you were right, and thanks for the thorough matrix. Before rewriting I re-verified independently in Chrome 149: all 12 placement×alignment cells with real geometry, `self-*` is byte-identical to the physical mapping in LTR and mirrors correctly in RTL (including with `flip-inline` fallbacks). My original "logical keywords don't work" conclusion was indeed from the plain family resolving against the containing block — the `self-*` family wasn't part of that test. Mea culpa.

Everything on the cleanup checklist is deleted: `getRtlJustifySelf`, the `isRtl` state, the show-time `getComputedStyle` read and style stamping, and the dual-ref stash + layout effect — `show()` is back to the bare feature-checked `showPopover()`. Net −582 lines. Kept `positioning: 'custom'` and `positionTryFallbacks`. Tests are now a 12-cell table of the emitted `self-*` strings plus a direction-independence check; the RTL Storybook story stays as the visual proof surface. Docs and changeset reworded to the pure-CSS story.

On the support window: I found no browserslist floor in the repo either, and Chrome 149 (one version behind your evidence) resolves `self-*` correctly, so I don't think there's a gap to defend against.

File: packages/core/src/Layer/useLayer.tsx (hunk 2 above)
Comment (AKnassa) [reply packages/core/src/Layer/useLayer.tsx:86]: Kept the `positioning` prop per your follow-up review. FWIW a `none` option on placement/alignment would need both to be `none` together to mean anything (a derived position-area needs both axes), so a single explicit switch seemed harder to misuse — but happy to revisit separately if you still prefer it.

File: packages/core/src/Carousel/Carousel.tsx (hunk 3 above)
Comment (AKnassa) [reply packages/core/src/Carousel/Carousel.tsx:402]: Carousel's overlay isn't expressible as a placement — it's an `anchor-size()` cover centered on the anchor (the buttons pin to the scroller's edges), so it stays `positioning: 'custom'`. With the `self-*` mapping there's no RTL styling to leak either way now.

**Comment by cixzhang** [comment]
Thanks this is looking good for the positioning. Noticing there's additional formatting updates here but they benign.

**Review by cixzhang — APPROVED** [review]
